[HWKINVENT-8] Versioning

metlos opened the pull request:
Work in progress on versioning.

Main principles:
- Nothing gets deleted, merely is marked as such (a new `eradicate()` method can truly delete stuff, but the backwards compatible `delete()` methods merely mark the elements as deleted from certain point in time onwards). 
- Entities are stored in 2 vertices - an identity vertex containing identifying info on the entity (its canonical path, type, etc) and state vertex that contains the rest of the entity properties as they exist at some point in time.
- The identity and state vertex are connected using an `__inState` edge that has `from` and `to` properties containing time stamps. Those define from and until when is given state valid.
- The current state has `to` set to `Long.MAX_VALUE` - a distant future.
- An entity is non-existent at some point in time if there is no identity vertex at all or there is no valid `__inState` edge for given point in time.
- All relationships have `from` and `to` properties, too, that define from and until when those relationships "exist".
- Relationship edges connect the identity vertices.
- All querying has been updated to take into account the `from` and `to` props.
- The "current" time is either left undefined, in which case we use `Instant.now()`, or is defined using `Inventory.at()` method.
- The current time is passed down to the backend as a `Discriminator` - the reasoning being that in future the discriminator might hold more than just time.

Still to do:
- [x] unit tests for `eradicate()` and "looking into the past"
- [x] /history REST
- [x] `?at=<time>` query param
- [x] create indices for time-related properties and update the pre-created schema with them
- [x] integration tests

metlos committed:
[HWKINVENT-8] Basic API for the versioning.

Nothing is implemented, this is just to get an idea of what it will
look like.

metlos committed:
Propagate Discriminator everywhere so that it can be used by the backend
to modify the queries based on "global" criteria like timerange

metlos committed:
[HWKINVENT-8] Most of stuff works.
The problem that needs most attention is the fact that after
deletion, the pre-commit no longer works, because it no longer can
find all the data because all the __to times have already been set
and therefore stuff marked for deletion is no longer queryable.

metlos committed:
Further fixes to how time discriminator is propagated, some fixes to
query generation, persisting a previously deleted entity works, etc.

There's just one failing integration test now...

metlos committed:
Trivial - formatting

metlos committed:
Fix RecurseFilter translation to Gremlin.

metlos committed:
Renamed "ResolvableToSingleWithRelationships" to "ResovableToSingleEntity"
and moved the history() method to it, because it only makes sense on
entities.

Actually implemented history().

Added *.Update.Builder.withDifference() method to compute an update from
2 entities - this is used in history() computation.

metlos committed:
Change the vertex existance check. A mere boolean was never gonna cut it.
Now, to find if a vertex represents an entity that exists at given point
in time, we just check that there is an appropriate __inState edge going
out of it - i.e. if an entity has a state at given point in time, it
exists, otherwise it doesn't.

metlos committed:
Trivial, make Titan buildable again by referencing the correct parent
version.

(It doesn't auto-upgrade it on releases anymore, because we don't build
Titan backend by default due to non-existance of a released version with
the code we need).

metlos committed:
Adding unit tests for eradicate and querying at specific times.

metlos committed:
Follow the docs and fail delete if there has been an update after
the designated time of deletion.

metlos committed:
Simplify Change class' generics.

metlos committed:
Added equals/hashCode to Action.Update and *.Update classes.
This will be needed for testing purposes in the later commits.

metlos committed:
Added support for Change JSON serialization.

metlos committed:
* Add support for ?at=<time> query param on all /tenant, /entity,
  /traversal, /sync and /bulk endpoints
* Add support for /history on /entity and /tenant
* Update Swagger docs with new REST params and endpoints
* Added stubs for to be implemented itests

metlos commented:
@jotak or @Jiri-Kremser could you please start reviewing this? This is a big PR because it's a big new feature :wink: Hopefully looking at it commit by commit will be manageable...

I still need to add integration tests for the new REST endpoints and query params but I'm hopeful that's gonna be relatively smooth.

I'll be offline tomorrow but will be available again on Thu and Fri..

jotak commented:
Big piece of work! :-) I haven't looked at the code yet, just reacting to the PR description (and thank you for that detailed description)

So if I understand, every modification of the vertex state, or of the relationships, will basically "close" the "to" time of the previous element and create a new element with opened "to". Which means the overall stored data will grow indefinitely, or is there any kind of TTL? Are we fine with that?
Since every query now uses "from" and "to", is the performance impact negligible?

jotak commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/Change.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override public int compareTo(Change<?, ?> o) {
+        int diff = time.compareTo(o.time);
```
Quick question here... which is more or less related to the code guidelines:
Since there's no null-check on "o": do you globally assume there is never ever "null" objects? In which case I'm fine with no null-check, we just have to be careful on code-review.
Or if there's a mix of sometimes nullable objects, sometimes non-nullable, then it might be appropriate to use checks like guava's "checkNotNull"...

metlos replied:
The contract of `compareTo()` is to throw an NPE if the parameter is null, if I am not mistaken. I guess I should add a null checks for the properties of the change to be non-null, though.

jotak replied:
Yes, I put my comment here because that's where I noticed potential NPE but the null-checks, if we decide to do them, should definitely be done in constructors or setters.

It's not a remark for this PR in particular, but more a general thought.

jotak commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/base/BaseData.java:
```diff
@@ -193,34 +191,36 @@ void preCreate(DataModificationChecks<BE> checks, DataEntity.Blueprint<R> bluepr
-            BE structuredData = tx.getRelationshipTarget(dataRel);
-
-            tx.deleteStructuredData(structuredData);
+// This is pre-versioning code - all this is only invoked during eradicate now...
```
It should probably be deleted before merging the PR, no?

metlos replied:
Yes, forgot to do that...

jotak commented on hawkular-inventory-impl-tinkerpop-parent/hawkular-inventory-impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/FilterVisitor.java:
```diff
@@ -85,7 +90,7 @@ public void visit(GraphTraversal<?, ?> query, Related related, QueryTranslationS
                     state.setInEdges(true);
                     state.setComingFrom(Direction.IN);
-                    query.inE(related.getRelationshipName());
+                    query.inE(related.getRelationshipName()).discriminate(discriminator);
```
Isn't it possible to "discriminate" the query globally once for all (and then, the discriminator would need to be propagated in the subsequent traversal). It would be safer than discriminating each case one by one, no?
Of course it would mean that the discriminator cannot be changed during a traversal, but if I understand the concept it's a kind of global setting, no?

metlos replied:
I think it could be done using some decorator strategy of the graph. I didn't actually know about that possibility until a couple of days ago (I thought that Tinkerpop uses graph strategies solely for query optimization but apparently it can also use them for adding "application logic" to the traversals).

That said I am not sure if such global thing would work, because I have a feeling we do a little bit of special handling here and there that does not follow otherwise uniform application of those checks. But it is a great point and we should definitely investigate the possibility.

jotak replied:
Well I didn't mean to do something really global, but just in the scope of the "query" object. What I don't know is, if you write "query.discriminate(disc).inE(related.getRelationshipName())" => is the discriminator propagated? In other words, is it equivalent to "query.inE(related.getRelationshipName()).discriminate(disc)" ?

metlos replied:
no, in here `discriminate()` is very similar to `has()`. It only applies to the current position in the traversal.

jotak commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/base/BasePreCommit.java:
```diff
@@ -534,7 +537,7 @@ boolean canBeIdentityRoot() {
             if (element == null || loadingTx != tx) {
                 loadingTx = tx;
-                Discriminator disc = Discriminator.time(Instant.now());
+                Discriminator disc = Discriminator.time(/*isDelete ? txStart :*/ Instant.now());
```
Can you elaborate, why in case of deletion we may have to use txStart, or not?

metlos replied:
This is a leftover from a wrong logic I put in place at some point during the impl. I should revert that...

jotak reviewed (approved):
I assume it's just reformatting

jotak commented on hawkular-inventory-itest/src/test/java/org/hawkular/inventory/rest/test/InventoryITest.java:
```diff
@@ -1482,6 +1482,31 @@ public void testTreeHash() throws Throwable {
 
+    @Test
+    public void testHistory_onEntity() throws Exception {
+        //TODO implement
```
Don't forget these :)

jotak commented:
(lol... forget the above comment "I assume it's just reformatting" ; it was about a particular commit, not the whole PR! :smile:)

metlos commented:
> So if I understand, every modification of the vertex state, or of the relationships, will basically "close" the "to" time of the previous element and create a new element with opened "to". Which means the overall stored data will grow indefinitely, or is there any kind of TTL? Are we fine with that?

There is currently no TTL and data grows indefinitely, yes. I assumed that for a v1 of the functionality this would be sufficient. But I agree that we should put a mechanism in place to delete or compact old entries. Would you mind creating a JIRA for that?

> Since every query now uses "from" and "to", is the performance impact negligible?

Some kind of these checks is going to always be necessary unless we want the whole graph be copy-on-write, which is I think even less practical.

This actually reminded me of one more thing that needs to be done - I forgot to add indices on these properties and update the schema precreation... Adding a bullet point for it in the description.

metlos commented on hawkular-inventory-impl-tinkerpop-parent/hawkular-inventory-impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/FilterVisitor.java:
```diff
@@ -85,7 +90,7 @@ public void visit(GraphTraversal<?, ?> query, Related related, QueryTranslationS
                     state.setInEdges(true);
                     state.setComingFrom(Direction.IN);
-                    query.inE(related.getRelationshipName());
+                    query.inE(related.getRelationshipName()).discriminate(discriminator);
```
This is a leftover from a wrong logic I put in place at some point during the impl. I should revert that...

metlos committed:
Distinguish between type's name, identity vertex's label and state
vertex's label. All come from the same enum, but have different
purposes. Also, make use of the vertex label in the relative path
query translation, which should improve performance of those queries.

metlos committed:
Add indices on "__from" and "__to" for edges. The state tables for
entity types are now precreated.

metlos committed:
Removing commented out code and factoring out the dead code...

metlos committed:
Don't use update during wire-up so that we don't create 2 versions of an
entity while creating it.

metlos committed:
Correctly restrict on changes having lasted less than a millisecond.

metlos committed:
Returning history bound to a certain time interval was mightily broken.

metlos committed:
Reapply PR #298 because of a messed up rebase...

Make the connection url, user and password configurable (#298)

also through sysprops and env vars.

metlos committed:
Make sure to not consider states ending at the millisecond of the
deletion as valid for consideration during the delete.

metlos committed:
Special handling of the queries on the current state.
These avoid checking the "__to" value against the "current time" but rather
use the Long.MAX_VALUE.
This, together with strict transactional behavior, can ensure that we
correctly handle a series of changes to an entity under the resolution of
out timestamps. If we compare with just Long.MAX_VALUE as "__to" time, we
can be sure we always query the current state. If we used the Instant.now()
for the current state, we could find entities that have been invalidated
by the previous updates within 1 millisecond of the current time.

At the same time, we are still able to use the Instant.now() as a valid
value for the current state when processing mutating operations, where we
need to have a "concrete" time to use.

metlos committed:
Trivial - remove debugging System.out and a trailing space...

metlos committed:
Fix query generation for Ids filter.

metlos committed:
Implement the integration tests for versioning in REST.

jotak commented:
LGTM (but a small one considering the amount of changes! :wink: )
